### ISSUE TITLE
New Windows Installer

### ISSUE THREAD

**Issue by karim** [sqlitebrowser/sqlitebrowser#1401]
I opened this issue to track all the efforts to migrate our current NSIS installer to MSI (Windows Installer).

I'm not talking about changing CPack NSIS to CPack WIX, but using [WiX Toolset](http://wixtoolset.org/) instead to create the installer from scratch.

Migrating to Windows Installer will give us a greater control over the installation and will give the user the best possible experience. Also I saw it being requested a lot. Hopefully we can do it before our next big release.

I need to collect some information first to be able to create the new installer the way everyone wants.

**Questions**

1. Do you want to install both 32-bit and 64-bit at the same time? *Current installer requires removing one before installing the other.*
2. Do you want to install the stable version (for daily use) **and** the nightly version (to test new features/bug fixes) at the same time?
3. Should all nightly builds be treated as the same product, that is, the new version replaces the old one, or should it be treated as a different product?
4. Should we provide [plugins](https://github.com/sqlitebrowser/sqlitebrowser/issues/579) as a different installer that can be installed later, or we bundle them together with DB4S? If bundled, should we allow the user to have the option to install them (Custom Install) or just install DB4S + Plugins (Full Install) automatically?

Instead of waiting I already answered these questions myself. :smile:

**Answers**

1. 32-bit and 64-bit are two different products and both can be installed together, even though I find it to be a useless option. It is only useful for 64-bit Windows and if you are on a 64-bit system why would you want to install 32-bit?
2. Stable and Nightly are treated as the same product. If you have stable version installed and tried to install a nightly version it will *upgrade* the stable to nightly.
3. All nightly are treated the same. Newer nightly version replaces the older one.
4. Everything is installed by default. There is no option to install the application without the plugins or install the plugins separately. 

Please feel free to comment if you didn't like any of these *default* settings so we can change them. I just made a PR (#1400) so we can test it and go from there.

Finally, all open issues related to this one should be closed and moved to here. Thank you! :smiley:

**Comment by chrisjlocke** [comment]
2. Not entirely sure this is a good idea?  The nightly, while it has some updates and fixes, could introduce a nasty bug - not sure the user would appreciate blowing away their previous version?  Guess its easy to redownload and reinstall again, but it could still be 'nasty'... ?  Just a thought...

**Comment by justinclift** [comment]
Good thinking. :smile:

It looks like we could add some suitable file extension associations to the installer too:

* https://stackoverflow.com/questions/2772452/how-to-associate-application-with-existing-file-types-using-wix-installer

My **personal** answers to the questions above are:

1. For my use case (investigating reported problems) it'd be useful to install both 32-bit and 64-bit ones at the same time.  That's not a "normal" use case though. :wink:
2. Being able to install both the **stable** (should be considered "known good") version + a nightly ("test out a new feature or fix") would be useful.  But only if it's not confusing.  If enabling tihs makes things potentially confusing for newbie users, I'd vote against this.
3. No real preference.  When debugging stuff it could be useful to have several nightlies installed in parallel somehow (maybe different dirs?).  But it's not critical and not something I'd probably commonly use anyway.
4. I'd just have everything in one installer, at least for now.  I could see that if someone wants to make their own add-on module (eg the [GIS stuff](https://github.com/sqlitebrowser/sqlitebrowser/wiki/SpatiaLite-on-Windows)) or something that places bits into an existing DB4S installation, that could be useful for people.

**Comment by justinclift** [comment]
@remram44 Do you reckon this would be of interest to people in the various Carpentries projects?

ie should someone jump on the new TopicBox server and point towards this discussion?

**Comment by justinclift** [comment]
Just pointed at this on Twitter too: https://twitter.com/sqlitebrowser/status/1002183986003800064

**Comment by karim** [comment]
> Not entirely sure this is a good idea? The nightly, while it has some updates and fixes, could introduce a nasty bug - not sure the user would appreciate blowing away their previous version?

True. But my point was, if you are using the stable version, reported a bug, found it fixed in the nightly, downloaded and installed it, then it should *upgrade* your current stable version. Your current settings and file associations are the same and you can continue to use them without a problem. After all, our stable version *is the latest nightly version* at some point. :smiley: 

Users can also stop at a specific version and do not upgrade if they are satisfied with the version they have now, or can downgrade to an older nightly/stable.

And later on, when we release a new stable version (with higher version than the latest nightly), they don't get *stuck* with a nightly version and get upgraded back to a stable version.

And let's not forget that this way will make it possible to have **four** entries in Add/Remove Programs. :smile: 

- DB Browser for SQLite (32-bit)
- DB Browser for SQLite (32-bit) [Nightly]
- DB Browser for SQLite (64-bit)
- DB Browser for SQLite (64-bit) [Nightly]

:scream:

**Comment by karim** [comment]
> It looks like we could add some suitable file extension associations to the installer too

Yeah. It is very powerful, but we should read the [tutorial](https://www.firegiant.com/wix/tutorial/) in order to get familiar with it.

> 1. For my use case (investigating reported problems) it'd be useful to install both 32-bit and 64-bit ones at the same time. That's not a "normal" use case though.

Personally, I don't like having 32-bit and 64-bit. It is good if you are using a portable version. You can run the 32-bit version if the system you are on is not 64-bit. But for an application you installed on your PC, why would you want to install a 32-bit version on a 64-bit PC? I don't find any use case for it.

However, I made it possible that both versions can be installed.

**Comment by karim** [comment]
> 3. No real preference. When debugging stuff it could be useful to have several nightlies installed in parallel somehow (maybe different dirs?). But it's not critical and not something I'd probably commonly use anyway.

Well, this might not be as easy as it seems and deserves its own comment. :smiley:

NSIS simply packs the files and throws them in a folder during installation. It is a simple extract + copy. It has no concept of versions. That made it possible to install multiple versions of the same application.

If you install it in the same location it will overwrite the old one, effectively upgrading it. If it is installed in a different location it will be installed as new. But you only get **one** entry in "Add/Remove Programs" and pointing to the last installed version. Uninstalling the application from "Add/Remove Programs" will uninstall the last one only.

MSI is different. It tracks each installed application by its GUID. Whenever you run a `.msi` file it will check this GUID. If the same GUID is found, it means the app is already installed, so it checks the version, if a higher version it will perform an upgrade, if lower version it will stop installing and say the app is installed with higher version already, if the same version it will show you the option to Repair/Configure it (add/remove features).

So if you install a version in `C:\This\Folder\` and then tried to install a different version of nightly in `D:\Other\Folder\` the Windows Installer will detect that both are the same application, check the version, if the one you are installing now is higher it will upgrade the application by removing the first folder and installing the app in the second folder. Only one version will exist.

As a workaround we can make each nightly build as a different product. This will make it possible to install each and every nightly version. We will lose the ability to upgrade though, since each version is a completely different product now.

Whenever a user installs a new version of nightly to fix a bug or get a new feature, he/she will have to go to "Add/Remove Programs" and uninstall the older version also.

**Comment by justinclift** [comment]
> As a workaround we can make each nightly build as a different product. This will make it possible to install each and every nightly version. We will lose the ability to upgrade though, since each version is a completely different product now.

Thanks Karim.  For this, I'm thinking it's more important for people to be able to do upgrades easily, as that's (I think) by far the more common use case.

So, lets go with the "same product" approach to enable easy upgrading.  I can figure out some workaround for the occasional time I need to install things in parallel or test an older version of bits. :smile:

**Comment by karim** [comment]
> I can figure out some workaround for the occasional time I need to install things in parallel or test an older version of bits.

Set `AllowSameVersionUpgrades="no"` attribute in `MajorUpgrade` element in `product.wxs` file then build the installer. Every new build will be a new product and can be installed with other builds built with this attribute set to `"no"`. When you are done testing, set it back to `"yes"`.

**Comment by justinclift** [comment]
Thanks. :smile:

**Comment by remram44** [comment]
@justinclift I have no need of multiple versions. The instructors are also unlikely to use different versions from what our learners would be using, since that would be confusing for them.

The best option would be everyone using the 'stable' version, but advocating the use of a nightly build could be possible if we need a specific fix or feature (but the only one we know helps for our lesson is #1349).

**Comment by justinclift** [comment]
Thanks @remram44. :smile:

**Comment by JReally** [comment]
May I suggest the process used by Lazarus installer... the end user has the option to create a "secondary" installation... see attached
![lazarusmultipleinstalls](https://user-images.githubusercontent.com/39771712/40811424-22f5fdf0-64ff-11e8-96c2-6689a8714467.png)

**Comment by justinclift** [comment]
Yep, all suggestions are welcome.

Just to confirm, the Lazarus installer you're meaning is this one yep?

&nbsp; &nbsp; http://wiki.freepascal.org/Lazarus_Installer

It looks like they're using an older generation technology as well (Inno Setup), however the concept might be workable if we dig in an investigate how it's done. :smile:

Priority wise, it'll probably be after we get the main bits working first though. :grinning:

**Comment by JReally** [comment]
I can't say about the installer at the link you listed (I would assume it is the one)
I just download Lazarus from this link http://www.lazarus-ide.org/
and get the option in the screen shot.

**Comment by justinclift** [comment]
Thanks @JReally. :smile:

**Comment by MKleusberg** [comment]
Just wondering, with the new Windows installer can we delete the cpack and/or NSIS bits at the end of the ```CMakeLists.txt``` file now or are they still needed? Same for the ```distri/winlaunch.bat``` file. :smile:

**Comment by karim** [comment]
> Just wondering, with the new Windows installer can we delete the cpack and/or NSIS bits at the end of the CMakeLists.txt file now or are they still needed? Same for the distri/winlaunch.bat file.

Yes, we can delete them. At the moment, we are creating both, .exe. and .msi, till we find no problems with the new installer. We need to make sure also that the new installer can do everything the same as the old one.

**Comment by MKleusberg** [comment]
Ah I see :smile: Thanks for the explanation!

**Comment by SilvioGrosso** [comment]
Hello everyone,

I am wondering whether, with this brand-new installer, it will be easier to package the "Math extension" [1] together with the upcoming next stable release ?
From what I have gathered, with the previous installer available for Windows users, this extension was difficult to make it working correctly and thus not available to test so far.

BTW, I suppose you are going to update your software with the Sqlite 3. 24.0 version [2] as well; but that's another different improvement, of course :-)

THANKS a lot for your work on SqliteBrowser ! 

[1] https://github.com/sqlitebrowser/sqlitebrowser/issues/1224
[2] https://www.sqlite.org/releaselog/3_24_0.html

**Comment by justinclift** [comment]
Oh, good point.  I keep on forgetting about the Math extension.  That shouldn't be too hard to include. :smile:

With SQLite 3.24.0, it's been in our nightly builds for a few days already. :grin:

**Comment by SilvioGrosso** [comment]
Thanks a lot for your reply.
Just decided to sign up as a patreon for your project : you deserve it, really ! :-)

**Comment by justinclift** [comment]
Wow!  Thanks heaps @SilvioGrosso.  That really does help. :+1:

**Comment by justinclift** [comment]
Heh Heh Heh

With your becoming a Patron, we've just reached our first goal!  Server costs look like they'll be covered now. Whoo! Hoooo! :grinning:

**Comment by justinclift** [comment]
This SQLite extension looks like it would be useful:

&nbsp; &nbsp; https://github.com/cldellow/sqlite-parquet-vtable

A fairly in-depth blog post explaining what it does, and what it's useful for too:

https://cldellow.com/2018/06/22/sqlite-parquet-vtable.html

Seems like it would make things a lot easier for people using 1GB+ SQLite data sets.

**Comment by justinclift** [comment]
Some useful follow up info on the sqlite-parquet-vtable extension:

&nbsp; &nbsp; https://news.ycombinator.com/item?id=17387166

**Comment by karim** [comment]
Just got reminded by another issue that this one is still open. Sorry about that, closing now. :smile:

**Comment by justinclift** [comment]
It's all good. :smile:
